### ISSUE TITLE
Reproducing SoccerMaster Calibration Metrics With Internal   SoccerMaster PnL Pipeline

### ISSUE THREAD

**Issue by aisatsanz** [haolinyang-hlyang/SoccerMaster#5]
We are trying to reproduce the SoccerMaster calibration metrics on
SN22-test-center using the public SoccerMaster code and pretrained
weights.

Repository/files used:
- codes/SoccerMaster/models/keypoints_detection.py
- codes/SoccerMaster/models/lines_detection.py
- codes/SoccerMaster/data/pnlcalib_utils/
- codes/SoccerMaster/sn_calibration/
- pretrained weights from HuggingFace:
    - SoccerMaster/backbone.pt
    - SoccerMaster/KeypointsDetection.pt
    - SoccerMaster/LinesDetection.pt

Our goal is to reproduce the reported 512x512 SoccerMaster numbers
from the paper/table.
What we implemented as a thin reproduction wrapper:
1. Load SoccerMaster config/checkpoint.
2. Build the SoccerMaster SoccerNetGSR_Detection dataloader.
3. Run the original SoccerMaster multitask model.
4. Decode keypoint heatmaps with:
from models.keypoints_detection import KeypointsDetectionMetrics
kp_decoder.get_keypoints_from_heatmap_batch_maxpool(
    kp_heatmap[:, :-1],
    scale=2,
    max_keypoints=1,
    min_keypoint_pixel_distance=15,
    return_scores=True,
)

5. Decode line heatmaps with:
from models.lines_detection import LinesDetectionMetrics
line_decoder.get_keypoints_from_heatmap_batch_maxpool_l(
    line_heatmap[:, :-1],
    scale=2,
    max_keypoints=2,
    min_keypoint_pixel_distance=10,
    return_scores=True,
)

6. Convert decoded peaks to PnL dictionaries using the internal
   SoccerMaster utilities:
from data.pnlcalib_utils.utils_heatmap import coords_to_dict,
complete_keypoints

kp_dict = coords_to_dict(kp_tensor, threshold=kp_thr,
ground_plane_only=True)[0]
line_dict = coords_to_dict(line_tensor, threshold=line_thr,
ground_plane_only=True)[0]

kp_dict, line_dict = complete_keypoints(
    kp_dict,
    line_dict,
    w=512,
    h=512,
    normalize=True,
)

7. Run internal SoccerMaster PnL calibration:

from data.pnlcalib_utils.utils_calib import FramebyFrameCalib

cam = FramebyFrameCalib(512, 512, denormalize=True)
cam.update(kp_dict, line_dict)

ground_result = cam.heuristic_voting_ground(refine_lines=True)
full_result = cam.heuristic_voting(refine=True, refine_lines=True)

8. Save camera JSON predictions into SoccerNet calibration zip
   format.

9. Evaluate with the internal SoccerMaster/SoccerNet scorer:

from sn_calibration.src.evalai_camera import evaluate

results = evaluate(
    gt_zip,
    prediction_zip,
    width=512,
    height=512,
)

For JaC@10 and JaC@20, we used the same internal
evaluate_camera_prediction, get_polylines, scale_points, and
mirror_labels functions, changing only the threshold.

Resolution/settings:

- Input size: 512x512
- Evaluation width/height: 512x512
- SoccerMaster config has:
    - AUG_RESIZE_SCALES: [512]
    - AUG_RANDOM_RESIZE: [512]
    - AUG_MAX_SIZE: 512
    - KEEP_ASPECT_RATIO: False

- Heatmap resolution: 256x256
- Decode scale: 2
- NUM_KEYPOINTS: 58
- NUM_LINES: 24

Thresholds tested:

- kp_thr=0.15, line_thr=0.15
- kp_thr=0.10, line_thr=0.10

Smoke results on the first 30 SN22-test-center frames:

kp=0.15, line=0.15:

CR      = 0.8667
JaC@5   = 0.5855
JaC@10  = 0.6616
JaC@20  = 0.6841
FS      = 0.5074

kp=0.10, line=0.10:

CR      = 0.9000
JaC@5   = 0.5154
JaC@10  = 0.5948
JaC@20  = 0.6281
FS      = 0.4639

The full SN22-test-center run is still running/queued, but the
early smoke results are substantially below the reported
SoccerMaster 512x512 numbers.

Reported table values we are trying to reproduce:

SN22-test-center, SoccerMaster/Ours* 512x512:
JaC@5  = 70.5
JaC@10 = 92.0
JaC@20 = 94.8
CR     = 99.4
FS     = 70.1

SN22-test-center, SoccerMaster/Ours† 512x512:
JaC@5  = 76.9
JaC@10 = 92.3
JaC@20 = 94.4
CR     = 98.6
FS     = 75.8

Questions:

1. Is the above inference pipeline the intended way to reproduce
   the SoccerMaster calibration metrics?

2. Should the PnL step use exactly:

FramebyFrameCalib(512, 512, denormalize=True)
complete_keypoints(..., normalize=True)
heuristic_voting_ground(refine_lines=True)
heuristic_voting(refine=True, refine_lines=True)

or is there a different calibration/refinement sequence used for
the reported table?

3. What confidence thresholds were used for keypoints and lines in
   the reported SoccerMaster calibration evaluation?
   We found conf_th=0.1 in the code, but 0.1/0.1 performs worse
   than 0.15/0.15 on our 30-frame smoke subset.

4. Are the reported SoccerMaster 512x512 metrics computed on
   square-resized images with KEEP_ASPECT_RATIO=False, or is there
   any inverse resize / original aspect-ratio handling before PnL
   or evaluation?

5. Is there an official script or command that reproduces the
   exact paper table from the released checkpoints?

6. Are the table numbers produced with only the released
   KeypointsDetection.pt and LinesDetection.pt, or do they also
   require the calibration/ReID pretrained models from sn-
   gamestate?

7. Does the reported method use only data/pnlcalib_utils from
   SoccerMaster, or does it rely on the external sn-gamestate/
   plugins/calibration/pnlcalib implementation?

8. Are non-planar goal classes included in the calibration
   evaluation output, or are they filtered/handled separately
   before scoring?

9. Is temporal context used at inference for calibration, and if
   yes, should camera predictions be smoothed or propagated across
   neighboring frames?

10. Could you provide the exact command/config used to generate
   the reported SN22-test-center and SN23-test calibration
   results?

**Comment by haolinyang-hlyang** [comment]
Hi,

Thanks for the detailed reproduction attempt. Below are the details I can still recall — hope they help.

I'll address your questions roughly in order:


1. Overall pipeline / Is this the intended way?

The pipeline you described is close but has some key differences from what I actually used. The main difference is the coordinate space and evaluation resolution. In my setup:

- Model inference is at 512x512 (as you're doing)
- But after decoding heatmaps, I scale all keypoint/line coordinates from 512x512 to 960x540
- FramebyFrameCalib is instantiated at 960x540, NOT 512x512
- Evaluation is also at 960x540 (the default of the evalai_camera.py script)

So the reported numbers in the paper were NOT produced by evaluating at 512x512. The "512x512" label in Table 4 refers to the model input resolution, but the calibration and evaluation happen at 960x540.


2. PnL calibration sequence

Here's the actual code I used (from my inference script), showing the complete post-processing pipeline after model forward pass:

    # Decode heatmaps (these functions are from PnLCalib:
    # https://github.com/mguti97/PnLCalib/blob/8c87391d6f4ea40c5e4d65e61529916c7a49ce62/utils/utils_heatmap.py)
    kp_coords = get_keypoints_from_heatmap_batch_maxpool(heatmaps[:, :-1, :, :])
    line_coords = get_keypoints_from_heatmap_batch_maxpool_l(heatmaps_l[:, :-1, :, :])
    kp_dicts = coords_to_dict(kp_coords, threshold=args.kp_th)
    lines_dicts = coords_to_dict(line_coords, threshold=args.line_th)

    # Process each image in the batch
    for i in range(b):
        kp_dict = kp_dicts[i]
        lines_dict = lines_dicts[i]

        # Complete keypoints from line intersections
        kp_dict, lines_dict = complete_keypoints(kp_dict, lines_dict, w=w, h=h)

        # Create calibration object at 960x540
        cam = FramebyFrameCalib(960, 540)

        # Scale keypoints and lines from 512x512 to 960x540
        scale_x = 960 / 512
        scale_y = 540 / 512

        for kp_id in kp_dict:
            kp_dict[kp_id]['x'] = int(kp_dict[kp_id]['x'] * scale_x)
            kp_dict[kp_id]['y'] = int(kp_dict[kp_id]['y'] * scale_y)

        for line_id in lines_dict:
            lines_dict[line_id]['x_1'] = int(lines_dict[line_id]['x_1'] * scale_x)
            lines_dict[line_id]['y_1'] = int(lines_dict[line_id]['y_1'] * scale_y)
            lines_dict[line_id]['x_2'] = int(lines_dict[line_id]['x_2'] * scale_x)
            lines_dict[line_id]['y_2'] = int(lines_dict[line_id]['y_2'] * scale_y)

        cam.update(kp_dict, lines_dict)
        final_params_dict = cam.heuristic_voting(refine_lines=True)

        if final_params_dict:
            if final_params_dict['rep_err'] <= args.max_reproj_err:
                # save camera params
                ...

Note:
- I did NOT use `denormalize=True` — coordinates are already in pixel space after scaling
- I did NOT use `normalize=True` in complete_keypoints
- I did NOT use `heuristic_voting_ground` — just `heuristic_voting(refine_lines=True)`, following PnLCalib's inference setting (https://github.com/mguti97/PnLCalib/blob/8c87391d6f4ea40c5e4d65e61529916c7a49ce62/scripts/inference_sn.py#L169)
- I did NOT use `ground_plane_only=True` in coords_to_dict
- I did NOT use `refine=True` in heuristic_voting (only `refine_lines=True`)
- Results are filtered by max reprojection error threshold (57 for SN22)


3. Confidence thresholds

The thresholds I used for SN22 are KP_TH=0.1611 and LINE_TH=0.3434 (from the sn-gamestate integration: https://github.com/SoccerNet/sn-gamestate/blob/1c958345067218297d221e45e1a6405f975f83e0/sn_gamestate/calibration/pnlcalib.py#L126-L127). For SN23 I used KP_TH=0.0712 and LINE_TH=0.2571 (PnLCalib's defaults). From what I recall, these values don't have a major impact on the final results, so this probably isn't the main source of your gap.


4. Aspect ratio / resize

Images are resized to square 512x512 with KEEP_ASPECT_RATIO=False (i.e., the image gets squished). After getting predictions in this 512x512 space, I scale coordinates back to 960x540 (the original SoccerNet resolution). There is no inverse resize on the image itself — only coordinate rescaling.


5. Official reproduction script / 10. Exact command

I no longer have access to the Python environment I used for the final experiments, and the code I have locally is not cleaned up for release. However, I believe the details above should be sufficient to reproduce the results. If you run into other issues, feel free to follow up on this issue.


6. Which checkpoints?

The "Ours*" (zero-shot) row uses the SoccerMaster pretrained backbone + jointly pretrained keypoint/line heads — this is the unified model after multi-task pretraining, without any further training on SoccerNet Calibration data.

The "Ours†" (fine-tuned) row: starting from the pretrained SoccerMaster weights, I fine-tuned the keypoints and lines separately — i.e., two independent fine-tuning runs, each producing its own checkpoint (backbone + corresponding head). Both the backbone and head are trained (no freezing), so the two resulting checkpoints have different backbone weights. At inference time, I load two separate models: one for keypoint prediction, one for line prediction — following PnLCalib's setting of using separate keypoint and line models.

Fine-tuning was done on SN22-train (center-camera subset only) for SN22 results, and SN23-train for SN23 results.

The backbone is siglip2-large-patch16-512 using the first Ls=16 spatial transformer blocks (out of 24 in total).


7. SoccerMaster code vs external sn-gamestate

I used the SoccerMaster codebase for multi-task pretraining (backbone + heads). Then for calibration fine-tuning and evaluation, I re-implemented the SoccerMaster backbone and head architectures within the PnLCalib codebase, loaded the pretrained weights, and ran fine-tuning and evaluation there. So the calibration utilities (FramebyFrameCalib, heuristic_voting, coords_to_dict, complete_keypoints, etc.) are all from PnLCalib (https://github.com/mguti97/PnLCalib), not from sn-gamestate. The code is mostly aligned with PnLCalib's pipeline.


8. Non-planar goal classes

Yes, goal posts and crossbars are included in the evaluation. The evaluation uses the full 3D pitch model (crossbar at z=2.44m, posts from z=0 to z=2.44m), and projection uses the full camera model. PnLCalib's line_world_coords_3D also includes goal elements.


9. Temporal context

No temporal context is used at inference for calibration. Each frame is processed independently — no smoothing or propagation across frames.

**Comment by aisatsanz** [comment]
| Split             | JaC@5 | JaC@10 | JaC@20 | CR | FS |
|-------------------|-------:|--------:|--------:|----:|----:|
| SN22-test-center  | 75.30% | 92.01% | 94.43% | 98.14% | 73.90% |
| SN23-test         | 64.94% | 80.21% | 84.37% | 71.03% | 46.12% |

**Comment by aisatsanz** [comment]
Thanks a lot for the detailed explanation, it was extremely helpful.

We updated our pipeline to match your description:
- decoded at 512×512
- scaled coordinates to 960×540
- FramebyFrameCalib(960, 540)
- complete_keypoints(..., normalize=False)
- coords_to_dict(..., ground_plane_only=False)
- heuristic_voting(refine_lines=True)
- no denormalize=True / no heuristic_voting_ground / no refine=True
- evaluation at 960×540

Using the released pretrained checkpoints, we now obtain the table I've sent previously. 
These numbers are now very close to the reported SN22 results, so your explanation resolved the reproduction issue.
Thanks again for taking the time to explain the evaluation pipeline and post-processing details.
